src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -293,6 +433,132 @@ internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projec
+                }
+
+                if (references.Contains(projectId))
+                {
+                    builder.Remove(project);
+                }
+            }
+
+            // Invalidate the transitive references from the previously-referenced project
+            builder.Remove(referencedProjectId);
+
+            return builder.ToImmutable();
+        }
+
+        internal ProjectDependencyGraph WithProjectRemoved(ProjectId projectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+
+            // Project ID set and direct forward references are trivially updated by removing the key corresponding to
+            // the project getting removed.
+            var projectIds = _projectIds.Remove(projectId);
+            var referencesMap = ComputeNewReferencesMapForRemovedProject(_lazyReverseReferencesMap, _referencesMap, projectId);
+
+            // The direct reverse references map is updated by removing the key for the project getting removed, and
+            // also updating any direct references to the removed project.
+            var reverseReferencesMap = _lazyReverseReferencesMap is object
+                ? ComputeNewReverseReferencesMapForRemovedProject(_referencesMap, _lazyReverseReferencesMap, projectId)
+                : null;
+            var transitiveReferencesMap = ComputeNewTransitiveReferencesMapForRemovedProject(_transitiveReferencesMap, projectId);
+            var reverseTransitiveReferencesMap = ComputeNewReverseTransitiveReferencesMapForRemovedProject(_reverseTransitiveReferencesMap, projectId);
+            return new ProjectDependencyGraph(
+                projectIds,
+                referencesMap,
+                reverseReferencesMap,
+                transitiveReferencesMap,
+                reverseTransitiveReferencesMap,
+                topologicallySortedProjects: default,
+                dependencySets: default);
+        }
+
+        internal ProjectDependencyGraph WithProjectReferenceRemoved(ProjectId projectId, ProjectId referencedProjectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+            Contract.ThrowIfFalse(_referencesMap[projectId].Contains(referencedProjectId));
+
+            var projectIds = _projectIds;
+            var referencesMap = _referencesMap.SetItem(projectId, _referencesMap[projectId].Remove(referencedProjectId));
+            var reverseReferencesMap = _lazyReverseReferencesMap is object
+                ? ComputeNewReverseReferencesMapForRemovedProjectReference(_lazyReverseReferencesMap, projectId, referencedProjectId)
+                : null;
+            var transitiveReferencesMap = ComputeNewTransitiveReferencesMapForRemovedProjectReference(_transitiveReferencesMap, projectId, referencedProjectId);
+            var reverseTransitiveReferencesMap = ComputeNewReverseTransitiveReferencesMapForRemovedProjectReference(_reverseTransitiveReferencesMap, projectId, referencedProjectId);
```

Comment:
these helpers are only called from one location.  i actually think it would be helpful to inline those funcitons into these methods.  That's because hte methods looks very similar in name.  i.e. `XxxFoRemovedProject` and `XxxForRemovedProjectReference`.  By pulling them into the only place they're called, i think it helps ensure clarity about which is used for which purpose.

Reply:
https://github.com/dotnet/roslyn/pull/40358/files#r369840076 applies here as well. This PR wasn't intended to restructure the high-level strategy used by this type.